docs/design/2018-08-29-new-planner.md

```diff
@@ -0,0 +1,411 @@
+  the same logical property.
+
+- **Physical Property**
+
+  Physical properties depend on the implementation algorithms, or physical
+  operators. Typical physical properties include **order** and **distribution**.
+
+## Proposal
+
+The new planner is composed of 2 phases: exploration and implementation. The
+basic idea comes from the volcano optimizer generator and the optimizations
+mentioned in the cascades project.
+
+### optimization phase: exploration
+
+A typical lifecycle of a SQL Query before this phase is:
+
+```
+SQL Query -> AST -> Logical Plan -> Expression Group
+```
+
+After building the expression group from the input logical plan, the exploration
+optimization phase begins, the target of this phase is to explore all the
+logically equivalent expressions by:
+
+1. Exploring all the equivalent group expressions of each group.
+2. Exploring all the potential groups.
+
+The group and group expression can be defined as:
+
+```go
+type StringSet = map[string]struct{}
+
+type Group struct {
+	equivalents  []*GroupExpr
+	fingerprints StringSet
+	explored     bool
+    ...
+}
+
+type GroupExpr struct {
+	exprNode LogicalPlan
+	children []*Group
+	explored bool
+    ...
+}
+```
+
+The difficult and complex part is how we continually apply some push-down and
+pull-up rules. A simple idea is to traverse the groups and group expressions
+twice: a top-down traverse and a bottom-up traverse. The push-down rules are
+applied during the top-down traverse, the pull-up rules are applied in the
+bottom-up traverse.
+
+But there might be some scenarios that certain push-down rule can also be
+triggered after the second bottom-up traverse. In order to explore all
+optimization possibilities, the traverse on the groups should not be stopped
+until there is no rule can be matched:
+
+```go
+func OnPhaseExploration(rootGroup *Group) error {
+	for !rootGroup.explored {
+		err := exploreGroup(rootGroup)
+	}
+}
+```
+
+The `explored` field in the `Group` and `GroupExpr` is used to avoid unnecessary
+traverse on the sub-tree of the groups and the group expressions. If a new group
+expression is inserted into a group, the new group expression and the group it
+belongs to is marked as un-explored, to enable further exploration on the new
+group expression and all the antecedent groups. The pseudo code to explore an
+expression group is:
+
+```go
+func exploreGroup(g *Group) {
+	if g.explored { return }
+
+	g.explored = true
+	for _, curExpr := range g.equivalents {
+		if curExpr.explored { continue }
+
+		// Explore child groups firstly.
+		curExpr.explored = true
+		for _, childGroup := range curExpr.children {
+			exploreGroup(childGroup)
+			curExpr.explored = curExpr.explored && childGroup.explored
+		}
+
+		eraseCur := findMoreEquiv(curExpr, g)
+		if eraseCur {
+			curGroup.erase(cur)
+		} else {
+			g.explored = g.explored && curExpr.explored
+		}
+	}
+}
+```
+
+The child of `GroupExpr` is `Group`, there are many candicate child expressions
```

Comment:
Please do not use a comma to combine two sentences. You can use:
The child of `GroupExpr` is **`Group`. There** are many **candidate** child expressions
or
The child of `GroupExpr` is `Group` **and** there are many **candidate** child expressions